keps/sig-windows/20190424-windows-cri-containerd.md

```diff
@@ -151,34 +155,284 @@ Operators could deploy additional RuntimeClasses with more granular control for
+#### Options for supporting multiple OS versions
+
+##### Option A - Use RuntimeClass per Windows OS version
+
+With process isolation, Windows containers must be constrained to only run on a matching Windows Server version node.
+
+[RuntimeClass Scheduling] has a few useful features that can make it easier to match a Windows container to a compatible Windows node. Here's an example of steps that could use RuntimeClass to avoid os version mismatches: 
+
+1. A RuntimeClass such as `WindowsServer1809` defined with [RuntimeClass Scheduling] `nodeSelector` for `kubernetes.io/os=windows` and `nodeSelector` for `beta.kubernetes.io/osversion=Windows1809` set.
+1. Nodes must be labeled with the `kubernetes.io/os=windows` (already there), as well as a new label for the OS version such as `beta.kubernetes.io/osversion=Windows1809`
+
+This would need to be repeated for each Windows version added to the cluster such as `windows-1809`, `windows-1903` and future releases. The cluster operator would need to ensure that nodes are added and scaled as needed for each Windows version independently.
+
+```yaml
+apiVersion: node.k8s.io/v1beta1
+kind: RuntimeClass
+metadata:
+  name: windows-1809-amd64
+handler: 'docker'  # only handler accepted when using dockershim
+scheduling:
+  nodeSelector:
+    kubernetes.io/os: 'windows'
+    kubernetes.io/arch: 'amd64'
+```
+
+A user doing a deployment must set an appropriate `runtimeClassName` for their Pod to be scheduled on a compatible Windows node.
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: mypod
+spec:
+  runtimeClassName: windows-1809-amd64
+  # ...
+```
+
+###### Migrating to Hyper-V from per-version RuntimeClass
+
+Once Hyper-V isolation is supported though, this causes a problem. There may be two different nodes that could run a given container built using Windows Server 1809:
+
+1. One running `Windows Server version 1809` supporting process isolation with `dockershim` or `containerd`
+    * Labels
+      * `kubernetes.io/os` = `windows`
+      * `beta.kubernetes.io/osversion` = `Windows1809`
+1. One running `Windows Server version 1903` supporting Hyper-V isolation with `containerd`
+    * Labels
+      * `kubernetes.io/os` = `windows`
+      * `beta.kubernetes.io/osversion` = `Windows1903`
+      * `capability.microsoft.com/isolation` = `hyperv`
+
+Using the `nodeSelector` above will only match node #1. Scheduling to node #2 would require defining a separate runtime class, and changing the `Pod` spec `runtimeClassName`. Changing the example above to run on Windows Server 1903 could look like this:
+
+```yaml
+apiVersion: node.k8s.io/v1beta1
+kind: RuntimeClass
+metadata:
+  name: windows-1809-amd64
+handler: 'windows-hyperv-1809'  # only handler accepted when using dockershim
+scheduling:
+  nodeSelector:
+    kubernetes.io/os: 'windows'
+    kubernetes.io/arch: 'amd64'
+    beta.kubernetes.io/osversion: '1903'
+    capability.microsoft.com/isolation: 'hyperv'
```

Comment:
add comment - that's a proposed new label not an existing one

Reply:
clarify - checking for the hyperv capability belongs in a label because not all cpu types / VM sizes may support Hyper-V

Reply:
this may need to be a more generic name like `supportsVirtualizationInstructions`